File: templates/problems/13_dp_1d/10_word_break/word_break_test.go
package word_break

import "testing"

func TestWordBreak(t *testing.T) {
	tests := []struct {
		name     string
		s        string
		wordDict []string
		want     bool
	}{
		{
			name:     "example leetcode",
			s:        "leetcode",
			wordDict: []string{"leet", "code"},
			want:     true,
		},
		{
			name:     "example applepenapple",
			s:        "applepenapple",
			wordDict: []string{"apple", "pen"},
			want:     true,
		},
		{
			name:     "cannot break",
			s:        "catsandog",
			wordDict: []string{"cats", "dog", "sand", "and", "cat"},
			want:     false,
		},
		{
			name:     "empty string",
			s:        "",
			wordDict: []string{"a"},
			want:     true,
		},
		{
			name:     "single char match",
			s:        "a",
			wordDict: []string{"a"},
			want:     true,
		},
		{
			name:     "single char no match",
			s:        "b",
			wordDict: []string{"a"},
			want:     false,
		},
		{
			name:     "overlapping words",
			s:        "cars",
			wordDict: []string{"car", "ca", "rs"},
			want:     true,
		},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			got := wordBreak(tt.s, tt.wordDict)
			if got != tt.want {
				t.Errorf("wordBreak(%q, %v) = %v, want %v", tt.s, tt.wordDict, got, tt.want)
			}
		})
	}
}

File: templates/problems/13_dp_1d/10_word_break/typescript/wordBreak.test.ts
import { describe, it, expect } from 'vitest';
import { wordBreak } from './wordBreak';

describe('WordBreak', () => {
    it('example leetcode', () => {
        expect(wordBreak('leetcode', ['leet', 'code'])).toBe(true);
    });

    it('example applepenapple', () => {
        expect(wordBreak('applepenapple', ['apple', 'pen'])).toBe(true);
    });

    it('cannot break', () => {
        expect(wordBreak('catsandog', ['cats', 'dog', 'sand', 'and', 'cat'])).toBe(false);
    });

    it('empty string', () => {
        expect(wordBreak('', ['a'])).toBe(true);
    });

    it('single char match', () => {
        expect(wordBreak('a', ['a'])).toBe(true);
    });

    it('single char no match', () => {
        expect(wordBreak('b', ['a'])).toBe(false);
    });

    it('overlapping words', () => {
        expect(wordBreak('cars', ['car', 'ca', 'rs'])).toBe(true);
    });
});
